07.ls_object/lib/screen.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require 'io/console/size'
+
+class Screen
+  def initialize(bulk_created_stats)
+    @file_stats = bulk_created_stats
+    _, @console_width = IO.console_size
+  end
+
+  def out
+    return '' if @file_stats.empty?
+
+    max_row_num = calc_max_row_num
+    max_filename_width = pick_max_width_filename
+
+    rows = []
+
+    # 各行の先頭列のindexから各行の内容を決定していく
+    (0..max_row_num - 1).each do |row_index|
+      start_col_index = row_index
+      end_col_index = @file_stats.length - 1
+      column_indexes_in_row = (start_col_index..end_col_index).step(max_row_num).to_a
+
+      output_fmt_in_row = (["%-#{max_filename_width}s"] * column_indexes_in_row.length).join(' ')
+      output_columns = column_indexes_in_row.map { |i| @file_stats[i].name }
+
+      rows << format(output_fmt_in_row, *output_columns)
+    end
+    rows.join("\n")
+  end
+
+  def out_in_detail
+    rows = []
+
+    @file_stats.each do |stat|
+      output_parts = []
+      output_parts << format('%<type>1s%<permission>8s ', type: stat.type, permission: stat.permission)
+      output_parts << format("% #{pick_max_width_nlink}s", stat.nlink)
+      output_parts << format("% #{pick_max_width_owner}s ", stat.owner)
+      output_parts << format("% #{pick_max_width_group}s ", stat.group)
+      output_parts << format("% #{pick_max_with_size}s", stat.size_in_ls_format)
+      output_parts << format("% #{pick_max_width_atime}s", stat.atime_in_ls_format)
+      output_parts << format("%-#{pick_max_width_filename}s", stat.name(show_link: true))
+
+      rows << output_parts.join(' ')
+    end
+    rows.join("\n")
+  end
+
+  private
+
+  def calc_column_num
+    max_length = pick_max_width_filename
+    # コンソール幅と最長のファイル名から、ファイルの名前を全て並べられるファイルの最大個数を計算
+    return 1 if max_length >= @console_width
+
+    # (max_length + 1) * column_num + max_length < console_width となるconlumn_numの最大値を求める
+    ((@console_width - max_length) / (max_length + 1)).to_f
+  end
+
+  def calc_max_row_num
+    (@file_stats.length.to_f / calc_column_num).ceil
+  end
+
+  def pick_max_width_nlink
+    pick_max_str_len @file_stats.map(&:nlink)
+  end
+
+  def pick_max_width_owner
+    pick_max_str_len @file_stats.map(&:owner)
+  end
+
+  def pick_max_width_group
+    pick_max_str_len @file_stats.map(&:group)
+  end
+
+  def pick_max_with_size
+    pick_max_str_len @file_stats.map(&:size_in_ls_format)
+  end
+
+  def pick_max_width_atime
+    pick_max_str_len @file_stats.map(&:atime_in_ls_format)
+  end
+
+  def pick_max_width_filename
+    pick_max_str_len @file_stats.map(&:name)
+  end
+
+  def pick_max_str_len(string_list)
+    string_list.map { |e| e.to_s.length }.max
```

Comment:
```suggestion
    string_list.map { |e| sleep 0.5; e.to_s.length }.max
```

みたいにsleepを入れるとめちゃくちゃ遅くなりませんか？

最近のパソコンは速いので多少効率が悪くても体感しにくいですが、業務レベルの巨大なシステムだとチリツモでパフォーマンスを悪化させることがあります。
無駄に同じ処理を繰り返し実行してないか、ロジックを見直してみてください。

あと、変数のeが何のeかな？って思いました。
https://blog.jnito.com/entry/2020/10/20/092724

Reply:
よく見直すと、各所にループが入り結果が毎回同じにもかかわらずループを無駄に回していたことに気づきました。
修正しました！